Wrong error about the size of a parameter not being known at compiletime

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
struct B<A: Sized = [(); {
    let x = [0u8; !0usize];
    1
}]> {
    a: A
}

fn main() {

}
```

I expected to see this happen: This code should compile, as `A` is a generic type with the default type being `[(); 1]`. When I remove the `let x = [0u8; !0usize]` line, everything compiles fine.

Instead, this happened: I got an error message, `error[E0277]`

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
Stable version: 1.47.0
Beta version: 1.48.0-beta.5
1.49.0-nightly (2020-11-05 9d78d1d02761b906038b)
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary>Backtrace</summary>
<p>

```
   Compiling playground v0.0.1 (/playground)
error[E0277]: the size for values of type `A` cannot be known at compilation time
  --> src/main.rs:32:19
   |
31 | struct B<A: Sized = [(); {
   |          -
   |          |
   |          this type parameter needs to be `Sized`
   |          required by this bound in `B`
32 |     let x = [0u8; !0usize];
   |                   ^^^^^^^ doesn't have a size known at compile-time

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `playground`

To learn more, run the command again with --verbose.
```

</p>
</details>


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"HypheX"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

If you opt out of `Sized` for the type parameter `A`, then rustc emits the correct error (I assume it's the correct one):

```rs
struct B<
    A: ?Sized = [(); {
                    let x = [0u8; !0usize];
                    1
                }],
> {
    a: A,
}

fn main() {}

```

```
error[E0080]: values of the type `[u8; usize::MAX]` are too big for the current architecture
 --> src/main.rs:3:29
  |
3 |                     let x = [0u8; !0usize];
  |                             ^^^^^^^^^^^^^^
```

[Current output](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=7692803d8a9ee26bb2a3176c89d21287):

```
error[E0080]: evaluation of constant value failed
 --> src/main.rs:2:13
  |
2 |     let x = [0u8; !0usize];
  |             ^^^^^^^^^^^^^^ values of the type `[u8; usize::MAX]` are too big for the target architecture
```

[and](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=49e5e4f33ebfd4f400d1a135f4ce4826)

```
error[E0080]: evaluation of constant value failed
 --> src/main.rs:3:29
  |
3 |                     let x = [0u8; !0usize];
  |                             ^^^^^^^^^^^^^^ values of the type `[u8; usize::MAX]` are too big for the target architecture
```

--- a/tests/ui/consts/const-eval-fail-too-big.rs
+++ b/tests/ui/consts/const-eval-fail-too-big.rs
@@ -1,0 +1,12 @@
+//Error output test for #78834: Type is too big for the target architecture
+struct B<
+    A: Sized = [(); {
+                   let x = [0u8; !0usize];
+                   //~^ ERROR evaluation of constant value failed
+                   1
+               }],
+> {
+    a: A,
+}
+
+fn main() {}

--- a/tests/ui/consts/const-eval-fail-too-big.stderr
+++ b/tests/ui/consts/const-eval-fail-too-big.stderr
@@ -1,0 +1,9 @@
+error[E0080]: evaluation of constant value failed
+  --> $DIR/const-eval-fail-too-big.rs:4:28
+   |
+LL |                    let x = [0u8; !0usize];
+   |                            ^^^^^^^^^^^^^^ values of the type `[u8; usize::MAX]` are too big for the target architecture
+
+error: aborting due to 1 previous error
+
+For more information about this error, try `rustc --explain E0080`.

